### DISCUSSION TITLE
Does child process fork() is same as spawn with explicit 'ipc' on Windows ?

### DISCUSSION THREAD

**Discussion by projektorius96** [nodejs/help#5011] [Q&A] [answered]
If the answer is negative what are the differences : **_is it true that fork() creates new Node.js V8 instance, whereas spawn does not_** ?

In terms of stdio variant I mentioned in the title : `stdio [null, null, null, 'ipc']` _whereas null for 'pipe'_

Any reference would be appreciated , thanks !

**Comment by cjihrig** [comment, marked as answer]
`spawn()` is a general purpose function for creating child processes. `spawn()` **can** be used to create Node.js child processes.

[`fork()` is built on top of `spawn()`](https://github.com/nodejs/node/blob/f02bdbf6a38093b3c3cf643288eaed610c24bc0f/lib/child_process.js#L170). It is only used for spawning Node.js child processes.

`fork()` does create an IPC channel. `spawn()` does not create an IPC channel by default, but can be configured to do so.

Check out the [`fork()` docs](https://nodejs.org/api/child_process.html#child_processforkmodulepath-args-options).

**Reply by projektorius96** [reply]
With reference to given fork docs it says (citing) :
> "Keep in mind that spawned Node.js child processes are independent of the parent with exception of the IPC communication channel that is established between the two. Each process has its own memory, with their own V8 instances. Because of the additional resource allocations required, spawning a large number of child Node.js processes is not recommended."

As far as I did understand, by default Node.js process is another process under OS process hierarchy (i.e. parent to child relationship), am I right ? @cjihrig  

If answer is more or less positive, does it mean that workers (threads) are kinda of "child_process" optimization within single Node.js process instance : rather than forking many independent Node.js V8 process instances i.e. instead of allocating more and more resources (RAM?) per each process,  instead we utitlise THREADS, that DO SHARE MEMORY ADDRESS SPACE **PER SINGLE PROCESS**, subsequently OS-wise context-switching is less expansive . Do I comprehend it correctly, at least conceptually ?  @cjihrig  

Thanks !

**Reply by cjihrig** [reply]
`fork()` creates a separate process at the operating system level. They are two completely separate Node.js processes that can communicate over IPC.

Worker threads are just that - threads. They are in the same Node.js process.

**Comment by projektorius96** [comment]
With reference to ["Shell scripting with Node.js"](https://exploringjs.com/nodejs-shell-scripting/downloads/nodejs-shell-scripting-book-preview.pdf) by [Dr. Axel Rauschmayer , @rauschma](https://github.com/rauschma) , cited : 

> ... **each worker is an V8 isolate** that has its own JavaScript heap but
shares its operating system heap with other threads.

Now getting back to your last sentence : "They (i.e. threads) are in the same Node.js process" .

Both cited above are quite mutually exclusive, don't you thinks so (politely) ?

I assume Axel Rauschmayer referred specifically to JavaScript V8-wise workers, whereas your i.e. @cjihrig last statement referred to Libuv-wise threads : if so I see no more contradiction herein as threads DO run within single context (≈isolate?) of Node.js process ? 

Can you confirm that "**Threads Node.js libuv implementation-wise**, rather than Workers V8-wise, **DO run within single context (≈isolate?) of Node.js process**" ?

Thank you .

**Reply by cjihrig** [reply]
The main thread and each Worker thread are separate threads in the same Node.js process. The main thread and each Worker thread has its own V8 isolate.

**Reply by projektorius96** [reply]
Last question, in terms of embedding V8 into C++ application, is there any definition for encapsulation of multiple isolates, say it "super-isolate ", "OS-wise namespace (like in Linux)" or some relevant ?

Thank you for patience & answers ! @cjihrig ♥

**Reply by cjihrig** [reply]
I have not heard of anything resembling a "super-isolate."

**Comment by ten9876** [comment]
Thanks for the excellent feedback, Jonathan! You're absolutely right.

## What we're changing

Our implementation was incorrectly offsetting the CW filter passband by the pitch frequency (e.g., `filter_lo=400, filter_hi=800` for pitch=600, width=400). The correct behavior is to center the filter on the carrier: `filter_lo=-200, filter_hi=200`.

The radio's internal BFO/demodulator handles making signals at the carrier audible at the sidetone pitch. The client should not apply any pitch offset to the filter position or the VFO marker.

## Fixes

1. **Filter presets**: CW/CWL filters now center on 0 Hz (the carrier), not on the pitch
2. **VFO marker**: Always drawn at the carrier frequency — no offset
3. **Click-to-tune**: Tunes carrier to clicked frequency — radio makes it audible at pitch
4. **Removed the Centered/Carrier toggle** from Radio Setup → Phone/CW — it was solving a problem that doesn't exist when the filter is correctly positioned

The result: marker, displayed frequency, and passband are all centered on the carrier. The passband naturally centers on the marker because `filter_lo = -width/2` and `filter_hi = +width/2`.

73 de KK7GWY
